opal/util/json/opal_json.h

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c)           Amazon.com, Inc. or its affiliates.
+ *                         All Rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+/** @file:
+ *
+ * Utility data structures and functions to parse JSON from a string or file.
+ * Note: The internal 3rd-party code should NOT be directly used. Implement the
+ * functionality here if necessary.
+ *
+ * Below is a trivial example to parse a JSON file with the content {"key": "val"}.
+
+int ret = OPAL_SUCCESS;
+opal_json_t *json = OBJ_NEW(opal_json_t), *item = OBJ_NEW(opal_json_t);
```

Comment:
I don't think I understand how this works. You have the internal type which contains an extra value pointer field, but here you only allocate an `opal_json_t` object. So, if you were to access these objects you will be generating invalid reads or segfaults. Now, the good thing is that you don't access these object, but instead you overwrite them in `opal_json_load`. This means this code would be leaking memory.

Reply:
I'm not sure about memory leak.

- `OBJ_NEW` allocates memory for the object - in this case it's `opal_json_internal_t` with the first member being `struct opal_json_t`, and invokes the constructor callbacks.
- `OBJ_RELEASE` first invokes the destructor callbacks - in this case it's `opal_json_internal_destruct` which calls `json_value_free` only once on the root json object. Then it will free the allocated `struct opal_json_internal_t` memory.

Could you please correct me if I'm wrong?

Reply:
`OBJ_NEW` indeed allocated memory for the object, such that `json` is a valid object. However, `opal_json_load_file` does not take an object it takes a pointer to an object, because internally it will create an object and return it to the caller (opal/util/json/opal_json.c lines 106 and 110). Thus, the pointer to the object you create here with `OBJ_NEW` will be overwritten, and your only reference to this object will be lost resulting in a memory leak.

Reply:
I see. Thank you! I updated the example.

Then I realized that it's odd that the caller does not actually call `OBJ_NEW` but still needs to do `OBJ_RELEASE`. Therefore I added `opal_json_free` that calls `OBJ_RELEASE` behind the scene.

So now we have `opal_json_load[_file]` to create the object and `opal_json_free` to free it.